src/shared/lib/resources/file.ts

```diff
@@ -38,46 +39,20 @@ export type FilePermissions<Id, UserType>
  * Parses a `FilePermissions` from a plain object.
  * Returns `null` if the parse fails or no permissions are supplied in the metadata.
  */
-// TODO this needs to be converted to a validation function using array combinators
-// TODO need to reduce the raw permissions into a "set" of permissions that prevents duplicates.
-export function parseFilePermissions(raw: any, parseOneUserType: (raw: string) => UserType | null): FileUploadMetadata {
-  const rawFilePermissions = isArray(raw) ? raw : [];
-  const filePermissions: Array<FilePermissions<Id, UserType>> = [];
-  rawFilePermissions.forEach(rawFilePermission => {
-    // TODO split this out into a validateFilePermission function
-    switch (get(rawFilePermission, 'tag')) {
-      case 'any':
-        filePermissions.push(adt('any' as const));
-        break;
-      case 'user':
-        if (isString(rawFilePermission.value)) {
-          filePermissions.push(adt('user' as const, rawFilePermission.value));
-        }
-        break;
-      case 'userType':
-        const userType = parseOneUserType(rawFilePermission.value);
-        if (userType) {
-          filePermissions.push(adt('userType' as const, userType));
-        }
-        break;
-      //TODO not an exhaustive switch statement... what if the user passes invalid permissions?
-    }
-  });
-  // TODO what if the user passes an empty array? Technically that is valid.
-  return filePermissions.length > 0 ? filePermissions : null;
-}
+export function parseFilePermissions(raw: any): FileUploadMetadata {
+  if (!isArray(raw)) {
+    return null;
+  }
 
-export function parseUserType(raw: string): UserType | null {
-  switch (raw.toUpperCase()) {
-    case 'GOV':
-      return UserType.Government;
-    case 'VENDOR':
-      return UserType.Vendor;
-    case 'ADMIN':
-      return UserType.Admin;
-    default:
-      return null;
+  // Validate each file permission
+  const validatedFilePermissions = validateArray(raw, validateFilePermission);
+  if (isValid(validatedFilePermissions)) {
+    // Reduce array of valid permissions into a unique set
+    const uniquePerms = Array.from(new Set(validatedFilePermissions.value.map(v => JSON.stringify(v))))
+      .map(v => JSON.parse(v) as FilePermissions<Id, UserType>);
+    return uniquePerms;
   }
```

Comment:
This should be moved into a separate function titled `validateFilePermissions` in `shared/lib/validation/file`. That function should include deduping the array of permissions. Then, `parseFilePermissions` can be a lot simpler.

Reply:
If we have time, at a later date, we could update the server adapter to support validating file metadata instead of just parsing them so API consumers are aware of why the file metadata is invalid.

Reply:
Sounds good - I added in proper invalid string values in the return from `validateFilePermission` in anticipation of this.